File: ai-concept/information_extractor.py
import json
from urllib.error import HTTPError, URLError
from urllib.request import Request, urlopen


OLLAMA_URL = "http://localhost:11434/api/generate"
MODEL = "llama3.2"
REQUIRED_FIELDS = (
    "customer",
    "product",
    "price",
    "order_date",
    "delivery_city",
    "delivery_date",
)


def call_ollama(text: str) -> str:
    prompt = f"""
Extract information from the text below.

Return one JSON object with exactly these fields:
- customer: string or null
- product: string or null
- price: number or null
- order_date: string or null
- delivery_city: string or null
- delivery_date: string or null

Rules:
- Use only information stated in the input.
- Use null when information is missing.
- Do not guess.
- Return JSON only, without Markdown or explanation.

Input text:
{text}
""".strip()

    body = json.dumps(
        {
            "model": MODEL,
            "prompt": prompt,
            "format": "json",
            "stream": False,
            "options": {"temperature": 0},
        }
    ).encode("utf-8")

    request = Request(
        OLLAMA_URL,
        data=body,
        headers={"Content-Type": "application/json"},
        method="POST",
    )

    with urlopen(request, timeout=120) as response:
        result = json.loads(response.read().decode("utf-8"))

    return result["response"]


def validate(data: object) -> dict:
    if not isinstance(data, dict):
        raise ValueError("The model did not return a JSON object.")

    missing_fields = [field for field in REQUIRED_FIELDS if field not in data]
    if missing_fields:
        raise ValueError(f"Missing fields: {', '.join(missing_fields)}")

    return {field: data[field] for field in REQUIRED_FIELDS}


def main() -> None:
    text = input("Enter an order description:\n").strip()

    if not text:
        print("Please provide an order description.")
        return

    try:
        model_output = call_ollama(text)
        extracted_data = validate(json.loads(model_output))
    except HTTPError as error:
        details = error.read().decode("utf-8", errors="replace")
        print(f"Ollama returned HTTP {error.code}: {details}")
        return
    except URLError:
        print("Could not connect to Ollama. Open the Ollama app and try again.")
        return
    except (json.JSONDecodeError, KeyError, ValueError) as error:
        print(f"The model returned invalid data: {error}")
        return

    print("\nValidated information:")
    print(json.dumps(extracted_data, indent=2, ensure_ascii=False))


if __name__ == "__main__":
    main()

File: ai-concept/information-extractor.mjs
import { createInterface } from "node:readline/promises";
import { stdin as input, stdout as output } from "node:process";

const OLLAMA_URL = "http://localhost:11434/api/generate";
const MODEL = "llama3.2";

const requiredFields = [
  "customer",
  "product",
  "price",
  "order_date",
  "delivery_city",
  "delivery_date",
];

async function callOllama(text) {
  const prompt = `
Extract information from the text below.

Return one JSON object with exactly these fields:
- customer: string or null
- product: string or null
- price: number or null
- order_date: string or null
- delivery_city: string or null
- delivery_date: string or null

Rules:
- Use only information stated in the input.
- Use null when information is missing.
- Do not guess.
- Return JSON only, without Markdown or explanation.

Input text:
${text}
  `.trim();

  const response = await fetch(OLLAMA_URL, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      model: MODEL,
      prompt,
      format: "json",
      stream: false,
      options: { temperature: 0 },
    }),
  });

  if (!response.ok) {
    const details = await response.text();
    throw new Error(`Ollama returned HTTP ${response.status}: ${details}`);
  }

  const result = await response.json();
  return result.response;
}

function validate(data) {
  if (data === null || Array.isArray(data) || typeof data !== "object") {
    throw new Error("The model did not return a JSON object.");
  }

  const missingFields = requiredFields.filter((field) => !(field in data));
  if (missingFields.length > 0) {
    throw new Error(`Missing fields: ${missingFields.join(", ")}`);
  }

  return Object.fromEntries(
    requiredFields.map((field) => [field, data[field]]),
  );
}

const readline = createInterface({ input, output });

try {
  const text = (await readline.question("Enter an order description:\n")).trim();

  if (!text) {
    console.log("Please provide an order description.");
    process.exitCode = 1;
  } else {
    const modelOutput = await callOllama(text);
    console.log(modelOutput);

    const extractedData = validate(JSON.parse(modelOutput));

    console.log("\nValidated information:");
    console.log(JSON.stringify(extractedData, null, 2));
  }
} catch (error) {
  console.error(`Could not extract information: ${error.message}`);
  process.exitCode = 1;
} finally {
  readline.close();
}
